### ISSUE TITLE
synapse fails to install with poetry 1.6.0 - "No module named 'synapse.synapse_rust'" at runtime

### ISSUE THREAD

**Issue by michaelkaye** [matrix-org/synapse#16147]
### Description

originally reported in: https://github.com/michaelkaye/setup-matrix-synapse/issues/95

This was last known working on 18 Aug at 17:00 UTC

No error seems to occur when poetry install is run.

### Steps to reproduce

Steps we do for install are written in the GHA action: https://github.com/michaelkaye/setup-matrix-synapse/blob/main/create.js#L18

See github action results for the error: https://github.com/michaelkaye/setup-matrix-synapse/actions/runs/5924305224

### Homeserver

testing server in GHA

### Synapse Version

main

### Installation Method

Other (please mention below)

### Database

n/a  - can't start synapse

### Workers

Single process

### Platform

Installed via michaelkaye/setup-matrix-synapse in a GHA worker

### Configuration

N/A - can't start synapse

### Relevant log output

```shell
/opt/pipx_bin/poetry run python -m synapse.app.homeserver --server-name localhost --config-path homeserver.yaml --generate-config --report-stats=no
Traceback (most recent call last):
  File "<frozen runpy>", line 189, in _run_module_as_main
  File "<frozen runpy>", line 112, in _get_module_details
  File "/home/runner/work/setup-matrix-synapse/setup-matrix-synapse/synapse/synapse/__init__.py", line 24, in <module>
    from synapse.util.rust import check_rust_lib_up_to_date
  File "/home/runner/work/setup-matrix-synapse/setup-matrix-synapse/synapse/synapse/util/rust.py", line 20, in <module>
    from synapse.synapse_rust import get_rust_file_digest
ModuleNotFoundError: No module named 'synapse.synapse_rust'
Error: The process '/opt/pipx_bin/poetry' failed with exit code 1
```

### Anything else that would be useful to know?

Installing via `pip` works fine - see alternate test cases in the github action.

**Comment by DMRobertson** [comment]
Can you try passing `-vv` to `poetry install` and seeing if that spews out any error context? It should generate a file called `synapse_rust.abi3.so`---suggest getting the action to do a `find . -name '*.so'` to confirm this exists in a sensible location. (For me this gets generated in the synapse directory (that containing e.g. `py.typed`).)

**Comment by DMRobertson** [comment]
> This was last known working on 18 Aug at 17:00 UTC

Hmm. Poetry did a release over the weekend. Do you pin the version of poetry that you use?

**Comment by DMRobertson** [comment]
Synapse uses 1.3.2, e.g.

https://github.com/matrix-org/synapse/blob/561d06b481176f61ed12f5a4723b127ff8624662/.github/workflows/tests.yml#L43

Though I would expect more recent versions to work as well.

**Comment by michaelkaye** [comment]
Thanks for the suggestions.

This is reproducible locally with the same error; i can't see the generated .so file mentioned above on my desktop.

We haven't pinned the version of poetry; locally i'm using poetry 1.6.0

Going to try:
 -  with the verbose flags on the GHA now.
  - with poetry pinned to 1.3.2 on the GHA now.

**Comment by michaelkaye** [comment]
`poetry -vv` doesn't output any more information: (perhaps we're actually hiding a lack of errors somewhere) https://github.com/michaelkaye/setup-matrix-synapse/actions/runs/5925091505/job/16063832535#step:4:69

`poetry 1.3.2` works fine: https://github.com/michaelkaye/setup-matrix-synapse/actions/runs/5925090308/job/16063828347

Going to pin to latest poetry before 1.6.0 to see if that helps, but maybe something in that latest release that's causing issues then.

**Comment by michaelkaye** [comment]
Even with -vvv the only extra messages I get in GHA are as follows, and checking for the .so file finds nothing:

```
 Installing the current project: matrix-synapse (1.90.0)
  - Building package matrix-synapse in editable mode
  - Falling back on using a setup.py
```

see https://github.com/michaelkaye/setup-matrix-synapse/actions/runs/5925479981/job/16064987664

Pinning poetry to 1.5.1 has solved my personal problem here; let me know if there's anything else debugging wise i can do to test

**Comment by clokep** [comment]
With poetry 1.6.1 I end up with a `.so` file in my build dir

```
$ find . -name '*.so'
./build/lib.macosx-13.3-x86_64-cpython-311/synapse/synapse_rust.abi3.so
```

If I move this into my Synapse directory it works fine: `mv ./build/lib.macosx-13.3-x86_64-cpython-311/synapse/synapse_rust.abi3.so synapse`

**Comment by clokep** [comment]
We're fairly certain a new release of poetry-core will fix this, probably due to python-poetry/poetry-core#633.

```diff
diff --git a/pyproject.toml b/pyproject.toml
index 5fb64479a1..23ef74885e 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -381,7 +381,7 @@ furo = ">=2022.12.7,<2024.0.0"
 # system changes.
 # We are happy to raise these upper bounds upon request,
 # provided we check that it's safe to do so (i.e. that CI passes).
-requires = ["poetry-core>=1.1.0,<=1.7.0", "setuptools_rust>=1.3,<=1.7.0"]
+requires = ["poetry-core @ git+https://github.com/python-poetry/poetry-core.git#main", "setuptools_rust>=1.3,<=1.7.0"]
 build-backend = "poetry.core.masonry.api"
 

```

**Comment by thebalaa** [comment]
This issue appears to be related to doing editable installs of this repo with `poetry install --extras all` or  `pip install -e .`

Manually copying the .so file from build/**/synapse_rust.abi3.so to ./synapse will usually resolve the issue.

tested with latest version of poetry 1.7.1

**Comment by thebalaa** [comment]
Again, poetry version is not and has never been the issue.

**Comment by michaelkaye** [comment]
If you can give me some specific instructions to follow (or a branch of synapse to try to install, etc etc) I'm happy to use the reproducible failure from the synapse GHA to prove success/failure of this using various poetry versions.

I might swing by the channel tomorrow?

**Comment by thebalaa** [comment]
I was mistaken this is a poetry issue after all: https://github.com/python-poetry/poetry/issues/8539

**Comment by dimbleby** [comment]
you will never know whether this is fixed in recent poetry-core, if you do not use recent poetry core
https://github.com/matrix-org/synapse/blob/825ac7e6a1a5a63cc1fd384e8cf4dd262d254965/pyproject.toml#L385

**Comment by clokep** [comment]
> you will never know whether this is fixed in recent poetry-core, if you do not use recent poetry core

We have been periodically tying updated poetry-core versions. We were really hoping 1.7.0 would fix it, but the issue seems to persist.

**Comment by dimbleby** [comment]
You knew in September that this was fixed in 1.8.0, I don't know when you forgot?

https://github.com/matrix-org/synapse/issues/16147#issuecomment-1741040730

**Comment by thebalaa** [comment]
This should fix it: https://github.com/matrix-org/synapse/pull/16702

**Comment by clokep** [comment]
> You knew in September that this was fixed in 1.8.0, I don't know when you forgot?
> 
> 
> 
> https://github.com/matrix-org/synapse/issues/16147#issuecomment-1741040730

I must have misquoted a version number, IIRC installing from git didn't work though and I didn't have further time to debug it. Maybe now that there's a release things will work properly.

**Comment by michaelkaye** [comment]
So I've poked at this off and on today:

Moving to poetry 1.7.1 for the `poetry` run,  (1.7.1 depends on poetry-core 1.8.1) on it's own, doesn't work.

We seem to need to both upgrade the poetry used to install and the one defined within synapse.

So poetry 1.7.1 + the patch in #16702 seems to work for getting this going.

See [workflows from setup-matrix-synapse](https://github.com/michaelkaye/setup-matrix-synapse/actions/runs/7036170569/job/19148201740) that are now completing with poetry versions > 1.5.x.

So basically I'd be in favour of incorporating this dependency bump.
